### PR TITLE
`Programming exercises`: Allow automatically graded exercises where all test cases give 0 points

### PR DESCRIPTION
**Pull request by Strohgelaender**
# This PR is based on #5994

#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.

### Motivation and Context
#6093

### Description
This PR removes the weight check that depends on the assessment type. Now, only negative weights lead to an error.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Programming Exercise

1. Navigate to the grading page of the exercise and set all tests to a weight of 0.
2. Check that saving works.
3. Check the solution:
a) It should show 100% if all tests are successful
b) It shows less then 100% if some test are not successful or if there are SCA deductions
5. Participate as a Student. You should always have 0 points, no matter how many tests passed.

#### Code Review
- [x] Review 1
- [ ] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

**Review by nikirovast — APPROVED** [review]
Tested on ts1, works as expected.

Just a suggestion: should we add a button to set all weights to some value, so that the user doesn't have to go all over them and set them manually? something like "copy to all weights" or so?

**Review by tobias-lippert — DISMISSED** [review]
Code LGTM

**Review by timo07k — DISMISSED** [review]
Everything works as expected. Tested on TS1.

**Review by PhilipWerz — DISMISSED** [review]
Saving after setting all test weights to 0 works. Everything else also works as expected. Tested on TS2.

**Review by JohannesStoehr — DISMISSED** [review]
Code looks good to me

**Review by simon-hng — DISMISSED** [review]
code

**Review by b-fein — APPROVED** [review]
Code changes look good. :+1:

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java [hunk 1]
```diff
@@ -195,9 +195,7 @@ private void checkCorrectBranchElseThrow(final ProgrammingExerciseParticipation
     /**
      * Retrieves the submission that is assigned to the specified participation and its commit hash matches the one from the build result.
      *
-     * @param participationId id of the participation
-     * @param buildResult     The build results
-     * @return The submission or empty no submissions exist
+     * @return The submission or empty if no submissions exist
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:198]:
```suggestion
     * @param participationId id of the participation
     * @param buildResult     The build results
     * @return The submission or empty if no submissions exist
```

**Review by b-fein — DISMISSED** [review]
Code re-approve.

**Review by laadvo — APPROVED** [review]
Exercises with all tests with weight 0 work as expected, the student remains at 0 points even after tests have been passed. Tested on ts1.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java [hunk 2]
```diff
@@ -874,8 +875,10 @@ private double calculatePointsForTestCase(final ProgrammingExerciseTestCase test
         final double testPoints;
         double exerciseMaxPoints = scoreCalculationData.exercise().getMaxPoints();
 
-        // A weight-sum of zero would let the solution show an error to the instructor as the solution score must be
-        // 100% of all reachable points. To prevent this, we weigh all test cases equally in such a case.
+        // In case of a weight-sum of zero the instructor must be able to distinguish between a working solution
+        // (all tests passed, 0 points) and a solution with test failures.
+        // Only the second case should show a warning while the first case is considered as 100%.
+        // Therefore, all test cases have equal weight in such a case.
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:881]: Should this really be equivalent to 100%? Instructors might be confused why they got 100% and if students get 100% as well.
We could also improve the warning in the client to check if there is at least one test that was not executed or that was not successful

**Review by JohannesStoehr — APPROVED** [review]
Code looks ok. Left one general thought about the necessity for 100%

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java (hunk 2 above)
Comment (Strohgelaender) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:881]: I'm also not completely happy with that solution, but that's out of the scope of this PR. We should come up with a solution that:
- Does not need to handle the solution results in a special way
- Does not show unnecessary warnings if the configuration is correct but the solution has a score < 100%. In this context, we could also check for cases where hidden tests were not executed (by Ares) and therefore lead to a score of < 100%, but all public tests pass.
- Allows the instructor to easily detect issues in the test cases or configuration.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java [hunk 3]
```diff
@@ -196,8 +196,8 @@ private void checkCorrectBranchElseThrow(final ProgrammingExerciseParticipation
      * Retrieves the submission that is assigned to the specified participation and its commit hash matches the one from the build result.
      *
      * @param participationId id of the participation
-     * @param buildResult     The build results
-     * @return The submission or empty no submissions exist
+     * @param buildResult     The build result
```
Comment (luisa1999) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:199]: Looks good

**Review by luisa1999 — APPROVED** [review]
Code is fine for me!

**Pull request merged**